be1-go/internal/handler/method/rumor/hrumor/buffer.go

```diff
@@ -0,0 +1,66 @@
+package hrumor
+
+import (
+	"fmt"
+	"popstellar/internal/errors"
+	"popstellar/internal/handler/method/rumor/mrumor"
+	"sync"
+	"time"
+)
+
+const bufferEntryLifeTime = 3 * time.Second
+
+type buffer struct {
+	sync.Mutex
+	values map[string]mrumor.Rumor
+}
+
+func newBuffer() *buffer {
+	return &buffer{
+		values: make(map[string]mrumor.Rumor),
+	}
+}
+
+func (b *buffer) insert(rumor mrumor.Rumor) error {
+	b.Lock()
+	defer b.Unlock()
+
+	ID := fmt.Sprintf("%s:%d", rumor.Params.SenderID, rumor.Params.RumorID)
+
+	_, ok := b.values[ID]
+	if ok {
+		return errors.NewDuplicateResourceError("rumor %s is already inside the buffer", ID)
+	}
+
+	b.values[ID] = rumor
+
+	go b.deleteWithDelay(ID)
+
+	return nil
+}
+
+func (b *buffer) deleteWithDelay(ID string) {
+	defer b.Unlock()
+
+	timer := time.NewTimer(bufferEntryLifeTime)
+
+	<-timer.C
+
+	b.Lock()
+	delete(b.values, ID)
```

Comment:
Having the defer unlock at the beginning works fine, but maybe having the (defer) unlock near/after the lock could be more readable.

Reply:
fixed